dag_in_context/src/typechecker.rs

```diff
@@ -1,11 +1,33 @@
 use std::{collections::HashMap, rc::Rc};
 
 use crate::{
-    ast::{base, emptyt, statet},
-    schema::{BaseType, BinaryOp, Constant, Expr, RcExpr, TernaryOp, TreeProgram, Type},
+    ast::{base, emptyt, infunc, inif, inloop, statet},
+    schema::{
+        Assumption, BaseType, BinaryOp, Constant, Expr, RcExpr, TernaryOp, TreeProgram, Type,
+    },
     tuplet,
 };
 
+pub(crate) struct TypeStack(Vec<Type>);
+
+impl TypeStack {
+    pub(crate) fn get(&self) -> &Type {
+        self.0.last().unwrap()
+    }
+
+    pub(crate) fn pushed(&self, ty: Type) -> TypeStack {
```

Comment:
why pushed and popped instead of push and pop?

Reply:
they don't mutate self